### PR TITLE
Network,Utility: remove usage of SemaphoreLocker

### PR DESCRIPTION
**Pull request by aarani**
Locking mutable shared variables has caused
weird bugs which were exteremly hard to find/detect,
this commit removes the usage of SemaphoreLocker in
the network layer of NOnion and replaces it by using
MailBoxes which allows implementation of multi-writer,
single-reader access pattern.

### CODE REVIEW

File: NOnion/Network/TorGuard.fs [hunk 1]
```diff
@@ -13,14 +13,98 @@ open NOnion
...
     let shutdownToken = new CancellationTokenSource()
 
     let mutable circuitsMap: Map<uint16, ITorCircuit> = Map.empty
     // Prevents two circuit setup happening at once (to prevent race condition on writing to CircuitIds list)
     let circuitSetupLock: obj = obj()
 
-    let sendLock = SemaphoreLocker()
+    let sendMailBoxProcessor(inbox: MailboxProcessor<GuardSendMessage>) =
+        async {
+            let innerSend circuitId (cellToSend: ICell) =
```
Comment (knocte) [inline NOnion/Network/TorGuard.fs:39]: no need to declare this nested function inside the async block, put it outside

File: NOnion/Network/TorGuard.fs [hunk 2]
```diff
@@ -13,14 +13,98 @@ open NOnion
...
+
 type TorGuard private (client: TcpClient, sslStream: SslStream) =
     let shutdownToken = new CancellationTokenSource()
 
     let mutable circuitsMap: Map<uint16, ITorCircuit> = Map.empty
     // Prevents two circuit setup happening at once (to prevent race condition on writing to CircuitIds list)
     let circuitSetupLock: obj = obj()
 
-    let sendLock = SemaphoreLocker()
+    let sendMailBoxProcessor(inbox: MailboxProcessor<GuardSendMessage>) =
```
Comment (knocte) [inline NOnion/Network/TorGuard.fs:37]: should be PascalCase

File: NOnion/Network/TorGuard.fs [hunk 3]
```diff
@@ -96,50 +180,22 @@ type TorGuard private (client: TcpClient, sslStream: SslStream) =
...
-                | exn ->
-                    match FSharpUtil.FindException<SocketException> exn with
-                    | Some socketExn ->
-                        return raise <| NOnionSocketException socketExn
-                    | None -> return raise exn
-            }
+    member __.Send (circuitId: uint16) (cellToSend: ICell) =
+        async {
+            let! sendResult =
+                sendMailBox.PostAndAsyncReply(fun _replyChannel ->
```
Comment (knocte) [inline NOnion/Network/TorGuard.fs:186]: @aarani this arg is used, so it shouldnt have an underscore

Comment (aarani) [reply NOnion/Network/TorGuard.fs]: Fixed in c5c9b5e18c3ff4052145c3a160806b6bbae1a2b2.

File: NOnion/Network/TorGuard.fs (hunk 1 above)
Comment (aarani) [reply NOnion/Network/TorGuard.fs]: Fixed in c5c9b5e18c3ff4052145c3a160806b6bbae1a2b2.

File: NOnion/NOnion.fsproj [hunk 4]
```diff
@@ -97,7 +98,7 @@
 
   <Target Name="CopyProjectReferencesToPackage" DependsOnTargets="ResolveReferences">
     <ItemGroup>
-      <BuildOutputInPackage Include="@(ReferenceCopyLocalPaths->WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
+      <BuildOutputInPackage Include="@(ReferenceCopyLocalPaths-&gt;WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
```
Comment (knocte) [inline NOnion/NOnion.fsproj:101]: this change is unrelated to this commit, please separate it to a new commit

File: NOnion/Network/TorCircuit.fs [hunk 5]
```diff
@@ -287,222 +305,434 @@ type TorCircuit
...
+                                tcs,
+                                _callback
+                            ) ->
+                            circuitState <-
+                                ReadyAsIntroductionPoint(
+                                    circuitId,
+                                    nodes,
+                                    _privateKey,
+                                    _publicKey,
+                                    _callback
```
Comment (knocte) [inline NOnion/Network/TorCircuit.fs:393]: @aarani these 3 elements are being used, so remove the underscore

File: NOnion/Network/TorCircuit.fs [hunk 6]
```diff
@@ -687,18 +878,325 @@ type TorCircuit
...
+                        introAuthPublicKey
+                        introEncPrivateKey
+                        introEncPublicKey
+                    |> TryExecuteAsyncAndReplyAsResult replyChannel
+
+            return! CircuitMailBoxProcessor inbox
+        }
+
+    let circuitOperationsMailBox =
+        MailboxProcessor.Start(CircuitMailBoxProcessor)
```
Comment (knocte) [inline NOnion/Network/TorCircuit.fs:974]: @aarani no need for parens here

File: NOnion/Network/TorCircuit.fs [hunk 7]
```diff
@@ -687,18 +878,325 @@ type TorCircuit
...
+        | RegisteringAsIntroductionPoint(circuitId, _, _, _, _, _)
+        | RegisteringAsRendezvousPoint(circuitId, _, _)
+        | WaitingForIntroduceAcknowledge(circuitId, _, _)
+        | WaitingForRendezvousRequest(circuitId, _, _, _, _, _, _)
+        | Ready(circuitId, _)
+        | ReadyAsIntroductionPoint(circuitId, _, _, _, _)
+        | ReadyAsRendezvousPoint(circuitId, _)
+        | Destroyed(circuitId, _)
+        | Truncated(circuitId, _) -> circuitId
+        | _ -> failwith "should not happen!"
```
Comment (knocte) [inline NOnion/Network/TorCircuit.fs:1006]: @aarani better message pls

File: NOnion/Network/TorStream.fs [hunk 8]
```diff
@@ -142,200 +139,377 @@ type TorStream(circuit: TorCircuit) =
...
+            | SendSendMe replyChannel ->
+                do! sendSendMe() |> TryExecuteAsyncAndReplyAsResult replyChannel
+
+            return! StreamControlMailBoxProcessor inbox
         }
 
-    member self.ConnectToDirectoryAsync() =
-        self.ConnectToDirectory() |> Async.StartAsTask
+    let streamControlMailBox =
+        MailboxProcessor.Start(StreamControlMailBoxProcessor)
```
Comment (knocte) [inline NOnion/Network/TorStream.fs:256]: @aarani no need for parens here

File: NOnion/Network/TorStream.fs [hunk 9]
```diff
@@ -142,200 +139,377 @@ type TorStream(circuit: TorCircuit) =
...
+
+            do!
+                safeReceive(buffer, offset, length)
+                |> TryExecuteAsyncAndReplyAsResult replyChannel
+
+            return! StreamReceiveMailBoxProcessor inbox
+        }
+
+    let streamReceiveMailBox =
+        MailboxProcessor.Start(StreamReceiveMailBoxProcessor)
```
Comment (knocte) [inline NOnion/Network/TorStream.fs:399]: @aarani no need for parens here

File: NOnion/Network/TorStream.fs [hunk 10]
```diff
@@ -142,200 +139,377 @@ type TorStream(circuit: TorCircuit) =
...
+
+
+    member __.SendData(data: array<byte>) =
+        async {
+            let! sendResult =
+                streamControlMailBox.PostAndAsyncReply(fun replyChannel ->
+                    StreamControlMessage.Send(data, replyChannel)
+                )
+
+            return UnwrapResult sendResult
```
Comment (knocte) [inline NOnion/Network/TorStream.fs:436]: @aarani Unwrap without message? let's fix the API of UnwrapResult to have an err msg

File: NOnion/Utility/MailboxUtil.fs [hunk 11]
```diff
@@ -0,0 +1,40 @@
+﻿namespace NOnion.Utility
+
+open System.Net.Sockets
+
+//FIXME: for some reason FSharpUtil is in NOnion namespace instead of NOnion.Utility
```
Comment (knocte) [inline NOnion/Utility/MailboxUtil.fs:5]: @aarani let's fix it in a 2nd commit

File: NOnion/Network/TorStream.fs (hunk 10 above)
Comment (aarani) [reply NOnion/Network/TorStream.fs:436]: No unwrap here works by just re-raising the exception that it contains which I think is enough as a msg, isn't it?

Comment (knocte) [reply NOnion/Network/TorStream.fs:436]: fair enough

File: NOnion/Utility/MailboxUtil.fs (hunk 11 above)
Comment (aarani) [reply NOnion/Utility/MailboxUtil.fs]: Fixed. open NOnion was still needed in that file for NOnionSocketException tho.

File: NOnion/Network/TorStream.fs (hunk 9 above)
Comment (aarani) [reply NOnion/Network/TorStream.fs]: Fixed

File: NOnion/Network/TorStream.fs (hunk 8 above)
Comment (aarani) [reply NOnion/Network/TorStream.fs]: Fixed

File: NOnion/Network/TorCircuit.fs (hunk 7 above)
Comment (aarani) [reply NOnion/Network/TorCircuit.fs]: Fixed

File: NOnion/Network/TorCircuit.fs (hunk 6 above)
Comment (aarani) [reply NOnion/Network/TorCircuit.fs]: Fixed

File: NOnion/Network/TorCircuit.fs [hunk 12]
```diff
@@ -287,222 +305,434 @@ type TorCircuit
...
+                            circuitState <- Truncated(circuitId, reason)
+                    | RelayData.RelayIntroduceAck ackMsg ->
+                        match circuitState with
+                        | WaitingForIntroduceAcknowledge(circuitId, nodes, tcs) ->
+                            circuitState <- Ready(circuitId, nodes)
 
-                                kpGen.Init(
-                                    Ed25519KeyGenerationParameters random
-                                )
+                            tcs.SetResult(ackMsg)
```
[the remark this answers is not in the document]
Comment (aarani) [reply NOnion/Network/TorCircuit.fs]: Fixed

File: NOnion/Network/TorCircuit.fs (hunk 5 above)
Comment (aarani) [reply NOnion/Network/TorCircuit.fs]: Fixed

File: NOnion/NOnion.fsproj (hunk 4 above)
Comment (knocte) [reply NOnion/NOnion.fsproj]: @aarani if you dont include this change in a 3rd commit, it will keep showing up.

Comment (aarani) [reply NOnion/NOnion.fsproj:102]: Done.

**Review by knocte — APPROVED** [review]
